Datastructure Features

vetschn opened the pull request:
This should close #27 and related issues.

vetschn committed:
Changed pre-commit configuration to reflect what's in the provided environment.yml

vetschn committed:
Half a refactor of the dsbsparse interface.

There are still some things to be done concerning the distributed transpose and smarter indexing logic for the block indexer.

vetschn committed:
Added a bunch of tests and test-skeletons.

These tests will fail right now, because we still need to figure out the higher dimension matrix stacks.

vetschn committed:
Fixed and renamed section_size retrieval.

This randomly returned negative section sizes somehow, no clue how we didn't spot this earlier. Added a test as well and renamed it to get_section_sizes because it's what it does.

vetschn committed:
Refactoring and documentation.

I extracted the block-view stride trick into a helper method and documented some more methods. I also added bare-bones block indexing logic & some checks. The zeros_like method can also be abstracted.

vincent-maillou commented:
As we are supporting `to_dense()` I think we should also support `to_sparse()` here are my reasons why:
- For the `Inv` method of the GF solvers. For the gpu support, we may want to densify the entire matrix on the GPU instead of the CPU (And I know that this dense solver is a bit of a debug thing, but I feel like if we keep it, it should be supported and implemented correctly)
- For debug/compatibility purposes: let say someone have developed a bit of code that operate on a sparse system matrix. Then we could plug and test very easily its idea in QuaTrEx
- I feel like it can happen very quickly that someone want to look at what the system matrix or the self energies looks like on a system that cannot  be exported with to_dense().

vetschn committed:
DSBCOO implementation

I added a first implementation of the coordinate format datastructure. Also refactored some stuff that the two current implementations share into a utils module. Of course also modified tests.

vetschn committed:
Support higher stack dimensions

I modified the code to facilitate higher stack dimensions, seems to work pretty nicely and passes the current test cases (had to modify them to allow higher stack dimensions). I redid some of the _block_view logic, now it's a bit nicer.

vetschn commented:
I managed to mostly handle the higher stack dimensions. However, there is a very weird bug in the distributed transposition somewhere that I haven't been able to pin down yet. Most of the time the `.dtranspose()` works beautifully, but for some seemingly random combinations of `stack_shape` and `comm.size` the data on certain ranks is wrong.

vetschn committed:
GPU aware implementation

I added a gpu_utils module to handle determining the array module and expose some useful routines. I also changed np.ndarray to ArrayLike/array_like in the type hints and the docstrings. Other than that, I just basically switched np to xp everywhere lol.

vetschn commented:
The GPU aware implementation is working as well now, i.e. all the datastructure test cases pass both with cupy and numpy backend. The data corruption problem mentioned above is not fixed yet and i need to add a couple more tests.

vetschn committed:
Some more updates to dtranspose

I solved the weird problem issue for the distributed transpose. Everything seems to work perfectly for the numpy backend atm (more tests needed...). However, I randomly get segmentations faults when running with cupy. To be investigated

vetschn commented:
Everything should work now, last steps are increasing test coverage.

One thing to note: Currently, as we have no GPU-aware MPI, `.dtranspose()` does not work with the `cupy` backend. Calling it when using `cupy` will raise segmentation faults (!)

vetschn commented on src/qttools/datastructures/dsbsparse.py:
```diff
@@ -1,244 +1,538 @@
+        # This does nothing if the data is already contiguous.
+        self._data = xp.ascontiguousarray(self._data)
+
+        synchronize_current_stream()
+        comm.Alltoall(MPI.IN_PLACE, self._data)
+
```
@vincent-maillou for the `cupy` backend, should we change this by just transferring the data to host for now? We can try to do more complicated stuff, like detecting if MPI is GPU aware later on... 

```suggestion
        synchronize_current_stream()
        comm.Alltoall(MPI.IN_PLACE, get_host(self._data))
```

vetschn committed:
Added checks for MPI GPU-awareness

I also added some more docs for the mpi_utils.

vetschn committed:
Small adjustments

Some small bugfixes i stumbled on when trying to implement tests.

vetschn committed:
Complete rehaul of datastructure test suite.

Basically implemented all the test from scratch and refactored everything into something nice. I also added another just command to determine test coverage and run the mpi tests separately if you want.

vetschn committed:
Remove original, half-baked, tests.

vetschn committed:
First attempt at fixing the testing workflow

vetschn committed:
Renamed get_num_elements_per_section to get_section_sizes.

This should be the last thing hindering the testing workflow from completing.

vetschn committed:
Updating test workflow

vetschn committed:
Try to fix github workflow

vetschn committed:
Remove essentially duplicate coverage report

vincent-maillou reviewed (approved):
It's awesome!
